### PR TITLE
[issue#39-Structure-for-training] Work in progress

### LINKED ISSUE #39: Обработка кода обучения и инференса модели для MNIST
В этом (нашем) репозитории нужно создать структуру для обучения и распознавания рукописных цифр. 

То есть, в результате выполнения этой задачи должно быть: 
- класс модели 
- класс инференса 
- ноутбук с инференсом 
- скрипт для обучения 

Обязательно везде должны быть комментарии. 
Эти файлы дожны быть разделены по отдельным файлам/папкам (тут уже на твоё усмотрение)

ВАЖНО! Данные, которые скачиваются из torch должны быть заигнорены, в репе такое хранить не надо (слишком жирно) 

Файл лучшей модели должен лежать в отдельной папке и загружен с помощью git lfs - https://git-lfs.github.com/ 
Файл модели должен иметь "говорящее" имя (= 

Тут уже с PR в конце (=

### PR DESCRIPTION
**Pull request by MarinaKupetskova**

### CODE REVIEW

File: Model_structure/model.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import torch.nn as nn
```
Comment (serykhelena) [inline Model_structure/model.py]: Этот файл должен быть внутри папки object_detection/mnist_model.py

File: Model_structure/config.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Configuration:
+    def __init__(self):
+        self.input_size = 784
+        self.hidden_size_1 = 200
+        self.hidden_size_2 = 150
+        self.hidden_size_3 = 100
+        self.hidden_size_4 = 80
```
Comment (serykhelena) [inline Model_structure/config.py]: давай конфиг для слоёв пока внутрь класса модели засунем, прям константами

File: Model_structure/config.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Configuration:
```
Comment (serykhelena) [inline Model_structure/config.py]: Этот класс не нужен пока

File: Model_structure/config.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    def __init__(self):
+        self.input_size = 784
+        self.hidden_size_1 = 200
+        self.hidden_size_2 = 150
+        self.hidden_size_3 = 100
+        self.hidden_size_4 = 80
+
+        self.output = 10
```
Comment (serykhelena) [inline Model_structure/config.py]: Это тоже внутрь класса модели

File: Model_structure/config.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        self.batch_size = 200
+        self.epoch = 100
+        self.lr_rate = 0.01
```
Comment (serykhelena) [inline Model_structure/config.py]: Размер батча, кол-во эпох и lr давай внутрь скрипта обучения

File: Model_structure/train.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
+import os
```
Comment (serykhelena) [inline Model_structure/train.py]: Этот код внутрь Scripts/mnist_train.py

File: Model_structure/train.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import torch
+import torch.nn as nn
+import torchvision
+from config import Configuration
+from model import MNIST
+from torch.optim.lr_scheduler import ReduceLROnPlateau
+from torchvision.transforms import transforms
+from tqdm import tqdm
+
+Config = Configuration()
```
Comment (serykhelena) [inline Model_structure/train.py]: вместо класса можно просто создать глобальный словарь, типа 
CONFIG = {
... 
} 

И в него указать размер батча, кол-во эпох и lr

File: Model_structure/train.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+)
+
+# Создаём лоядеры данных.
+# так как модель ожидает данные в определённой форме
+train_dataloader = torch.utils.data.DataLoader(
+    dataset=train_data, batch_size=Config.batch_size, shuffle=True
+)
+valid_dataloader = torch.utils.data.DataLoader(
+    dataset=valid_data, batch_size=Config.batch_size, shuffle=False
+)
```
Comment (serykhelena) [inline Model_structure/train.py]: получение выборок можно обернуть в отдельную функцию, типа get_loaders

File: Model_structure/train.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+Config = Configuration()
+
+# загружаем обучающую выборку
+train_data = torchvision.datasets.MNIST(
+    "mnist_content", train=True, transform=transforms.ToTensor(), download=True
+)
+
+# разделяем обучающую выборку на обучающую и валидационную выборки
+# 70% для обучения, 30% для валидации
+train_size = int(len(train_data) * 0.7)
```
Comment (serykhelena) [inline Model_structure/train.py]: пропорцию для разделения можно тоже в конфиг вынести

File: Model_structure/train.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+    output=Config.output,
+)
+# сразу отправить модель на устройство
+model = model.to(DEVICE)
+
+
+if __name__ == "__main__":
+
+    # функция потерь
+    criterion = nn.CrossEntropyLoss()
```
Comment (serykhelena) [inline Model_structure/train.py]: Этот код вынести в отдельную функцию main(), а внутри `if __name__ == "__main__":` будет её вызов `main()` 
Это одна из хороших практик = шаблон скрипта

File: Model_structure/train.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+model = MNIST(
+    input_size=Config.input_size,
+    hidden_size1=Config.hidden_size_1,
+    hidden_size2=Config.hidden_size_2,
+    hidden_size3=Config.hidden_size_3,
+    hidden_size=Config.hidden_size_4,
+    output=Config.output,
+)
+# сразу отправить модель на устройство
+model = model.to(DEVICE)
```
Comment (serykhelena) [inline Model_structure/train.py]: код с моделью тоже внутри функции main()

File: Model_structure/train.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    optimizer = torch.optim.Adam(model.parameters(), lr=Config.lr_rate)
+    # создаём сущность, которая автоматически уменьшит шаг обучения в случае,
+    # когда функция потерь перестанет уменьшаться в течение N эпох (patience)
+    scheduler = ReduceLROnPlateau(
+        optimizer, mode="min", patience=10, min_lr=1e-8, verbose=True
+    )
+
+    NUM_EPOCHS = Config.epoch
+
+    checkpoint_dpath = os.path.join(CURRENT_DIR, "mnist_checkpoints")
```
Comment (serykhelena) [inline Model_structure/train.py]: давай сделаем отдельную папку в корне проекта checkpoints, а внутри уже mnist_checkpoints. Папку checkpoints в гит-игнор

**Review by serykhelena — CHANGES_REQUESTED** [review]

File: Scripts/mnist_train.py [hunk 13]
```diff
@@ -1,86 +1,86 @@
 import os
+import sys
 
 CURRENT_DIR = os.path.dirname(os.path.abspath("__file__"))
+sys.path.append(CURRENT_DIR)
```
Comment (serykhelena) [inline Scripts/mnist_train.py]: сюда нужно ROOT_DIR

File: Scripts/mnist_train.py [hunk 14]
```diff
@@ -1,86 +1,86 @@
...
 from tqdm import tqdm
 
-Config = Configuration()
+from object_detection.mnist_model import MNIST
+
+CONFIG = {"batch_size": 200, "epoch": 100, "lr_rate": 0.01, "propotion": 0.7}
+
 
 # загружаем обучающую выборку
 train_data = torchvision.datasets.MNIST(
```
Comment (serykhelena) [inline Scripts/mnist_train.py]: загрузку данных тоже в фукнцию getLoaders

File: Scripts/mnist_train.py [hunk 15]
```diff
@@ -1,86 +1,86 @@
...
-    hidden_size1=Config.hidden_size_1,
-    hidden_size2=Config.hidden_size_2,
-    hidden_size3=Config.hidden_size_3,
-    hidden_size=Config.hidden_size_4,
-    output=Config.output,
-)
-# сразу отправить модель на устройство
-model = model.to(DEVICE)
 
+def getLoaders():
```
Comment (serykhelena) [inline Scripts/mnist_train.py]: в питоне по PEP принято называть функции в snake-case стиле, т.е. здесь должно быть get_loaders()

File: Scripts/mnist_train.py [hunk 16]
```diff
@@ -1,86 +1,86 @@
...
-    best_metric = 0
 
     for epoch in range(NUM_EPOCHS):
         print(f"--- Epoch {epoch} ---")
         for phase in ["train", "val"]:
             epoch_loss = []
             if phase == "train":
                 model.train()
-                loader = train_dataloader
+                loader = getLoaders()[0]  # train_dataloader
```
Comment (serykhelena) [inline Scripts/mnist_train.py]: лучше отдельно до цикла определить train_dataloader, test_dataloader = get_loaders()

**Review by serykhelena — CHANGES_REQUESTED** [review]

File: Scripts/mnist_train.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
+import os
+import sys
+
+CURRENT_DIR = os.path.dirname(os.path.abspath("__file__"))
+ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
```
Comment (serykhelena) [inline Scripts/mnist_train.py]: здесь можно использовать CURRENT_DIR, вместо `os.path.dirname(__file__)`

File: Scripts/mnist_train.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+                model.train()
+                loader = train_dataloader
+            else:
+                model.eval()
+                loader = valid_dataloader
+
+            for images, labels in tqdm(
+                loader, desc=f"{phase.upper()} Processing"
+            ):
+                images = images.reshape(-1, 28 * 28)
```
Comment (serykhelena) [inline Scripts/mnist_train.py:96]: images тоже надо `.to(DEVICE)` 
labels тоже `.to(DEVICE)`  

Иначе на GPU не будет работать, нужно, чтобы всё было на одном устройстве

**Review by serykhelena — CHANGES_REQUESTED** [review]

**Comment by serykhelena** [comment]
В ноутбуке инференса, тестовый данные взять из папки mnist_content (которая была создана при обучении). Ещё одну папку внутри mnist_reserarch делать не надо

**Comment by serykhelena** [comment]
В ноутбуке инференса добавить DEVICE, и добавить везде где нужно `.to(DEVICE)` (т.е. для модели, тестовой картинки)

File: Scripts/mnist_train.py [hunk 19]
```diff
@@ -94,6 +94,9 @@ def main():
                 loader, desc=f"{phase.upper()} Processing"
             ):
                 images = images.reshape(-1, 28 * 28)
+                images = images.to(DEVICE)
+
+                labels = labels.to(DEVICE)
 
                 optimizer.zero_grad()
                 with torch.set_grad_enabled(phase == "train"):
```
Comment (serykhelena) [inline Scripts/mnist_train.py:102]: Нужно убрать лишний print - на 103 строке

**Review by serykhelena — APPROVED** [review]
